native/src/snappyclient/cpp/impl/ClientService.cpp

```diff
@@ -216,14 +221,36 @@ void ClientService::handleUnknownException(const char* op) {
 }
 
 void ClientService::handleSnappyException(const thrift::SnappyException& se) {
+  FailoverStatus status = NetConnection::getFailoverStatus(se.exceptionData.sqlState,
+                                          se.exceptionData.errorCode,se);
+  if(status== FailoverStatus::NONE){
   throw GET_SQLEXCEPTION(se);
+  }else if(status== FailoverStatus::RETRY){
+    // need to do failover to new server, so get the next one
+    //close the connection explicitly since server may still be alive
+    close();
+    std::set<thrift::HostAddress> failedServers;
+    updateFailedServersForCurrent(failedServers,true,nullptr);
+    thrift::HostAddress hostAddr= m_connHosts.at(0);
+    openConnection(hostAddr,failedServers);
+  }
 }
 
 void ClientService::handleTTransportException(const char* op,
     const TTransportException& tte) {
-  checkConnection(op);
-
-  throwSQLExceptionForNodeFailure(op, tte);
+  if (this->m_loadBalance){
+        // need to do failover to new server, so get the next one
+        //close the connection explicitly since server may still be alive
+        close();
+        std::set<thrift::HostAddress> failedServers;
+        updateFailedServersForCurrent(failedServers,true,nullptr);
+        thrift::HostAddress hostAddr= m_connHosts.at(0);
```

Comment:
What happens if at(0) returns null as in no element is left for failover. Does openConnection handles that? It looks like its a place holder which is being filled later once the connection is established to the new server. Please check.

Reply:
took care of "if at(0) return null"